File: lessons/dsa/hash-lookup-as-a-trade/examples/lookup-costs.test.ts
import { expect, test } from "vitest";

import {
  hashSlot,
  keysExaminedByHash,
  keysExaminedByScan,
} from "./lookup-costs.js";

const KEYS = ["cat", "dog", "emu", "fox", "owl", "bat", "ant", "cow"];

test("a scan examines every key before the wanted one", () => {
  expect(keysExaminedByScan(KEYS, "owl")).toBe(5);
});

test("a scan examines all keys when the wanted one is absent", () => {
  expect(keysExaminedByScan(KEYS, "yak")).toBe(KEYS.length);
});

test("hashing examines only the wanted slot", () => {
  expect(keysExaminedByHash(KEYS, "owl", 16)).toBeLessThan(3);
});

test("hashing stays cheap as the collection grows", () => {
  const larger = [
    ...KEYS,
    ...Array.from({ length: 200 }, (_, index) => `key${index}`),
  ];

  expect({
    hashed: keysExaminedByHash(larger, "owl", 512) < 3,
    scanned: keysExaminedByScan(larger, "owl") > 3,
  }).toEqual({ hashed: true, scanned: true });
});

test("a slot is stable for a key and inside the table", () => {
  expect({
    stable: hashSlot("cat", 6) === hashSlot("cat", 6),
    inside: KEYS.every((key) => hashSlot(key, 6) >= 0 && hashSlot(key, 6) < 6),
  }).toEqual({ stable: true, inside: true });
});

test("an absent key still examines only its own slot", () => {
  expect(keysExaminedByHash(KEYS, "yak", 16)).toBeLessThan(3);
});

File: lessons/dsa/hash-lookup-as-a-trade/examples/lookup_costs_test.go
package examples

import (
	"fmt"
	"testing"
)

var keys = []string{"cat", "dog", "emu", "fox", "owl", "bat", "ant", "cow"}

func TestScanReachesTheWantedKey(t *testing.T) {
	if got := KeysExaminedByScan(keys, "owl"); got != 5 {
		t.Fatalf("expected 5 keys examined, got %d", got)
	}
}

func TestScanExaminesAllWhenAbsent(t *testing.T) {
	if got := KeysExaminedByScan(keys, "yak"); got != len(keys) {
		t.Fatalf("expected %d keys examined, got %d", len(keys), got)
	}
}

func TestHashExaminesOnlyItsSlot(t *testing.T) {
	if got := KeysExaminedByHash(keys, "owl", 16); got >= 3 {
		t.Fatalf("expected fewer than 3 keys examined, got %d", got)
	}
}

func TestHashStaysCheapAsTheCollectionGrows(t *testing.T) {
	larger := append([]string{}, keys...)
	for index := 0; index < 200; index++ {
		larger = append(larger, fmt.Sprintf("key%d", index))
	}

	if got := KeysExaminedByHash(larger, "owl", 512); got >= 3 {
		t.Fatalf("expected hashing to stay cheap, got %d", got)
	}
	if got := KeysExaminedByScan(larger, "owl"); got <= 3 {
		t.Fatalf("expected scanning to grow, got %d", got)
	}
}

func TestASlotIsStableAndInsideTheTable(t *testing.T) {
	if HashSlot("cat", 6) != HashSlot("cat", 6) {
		t.Fatal("expected a key to hash to the same slot twice")
	}
	for _, key := range keys {
		if slot := HashSlot(key, 6); slot < 0 || slot >= 6 {
			t.Fatalf("slot %d for %q is outside the table", slot, key)
		}
	}
}

func TestAnAbsentKeyExaminesOnlyItsSlot(t *testing.T) {
	if got := KeysExaminedByHash(keys, "yak", 16); got >= 3 {
		t.Fatalf("expected fewer than 3 keys examined, got %d", got)
	}
}
